### PR TITLE
Implement fetchSize functionality in FetchProjector.

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java [hunk 1]
```diff
@@ -101,7 +101,7 @@ public void nodeOperation(final NodeFetchRequest request,
             statsTables.operationStarted(request.fetchPhaseId(), request.jobId(), "fetch");
 
             // nothing to fetch, just close
-            if (request.toFetch() == null) {
+            if (request.toFetch() == null && request.isCloseContext()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java]: does this mean we send requests where there is nothing to fetch and where isCloseContext is false? Isn't that a bit of a waste ?

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java [hunk 2]
```diff
@@ -347,7 +347,6 @@ private boolean emitRow(Row row) {
             combinedRow.outerRow = left.lastRow;
             combinedRow.innerRow = row;
             boolean wantsMore = downstream.setNextRow(combinedRow);
-            assert downstreamWantsMore : "shouldn't emit rows if downstream doesn't need anymore";
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: is it possible to re-add the assertion or does this case still happen?

File: sql/src/main/java/io/crate/operation/merge/IteratorPageDownstream.java [hunk 3]
```diff
@@ -117,6 +117,10 @@ private void processBuckets(Iterator<Row> iterator, PageConsumeListener listener
                 return;
             }
         }
+        if (topRowUpstream.shouldPause()) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/merge/IteratorPageDownstream.java]: why is this required here? This is already done directly after the setNextRow call?

File: sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java (hunk 2 above)
Comment (matriv) [inline sql/src/main/java/io/crate/operation/join/NestedLoopOperation.java]: This case still happens.

File: sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+
+    private FetchProjector prepareFetchProjector(int fetchSize,
+                                                 CollectingRowReceiver rowReceiver,
+                                                 TransportFetchOperation transportFetchOperation,
+                                                 RowUpstream upstream) {
+        FetchProjector pipe =
+            new FetchProjector(
+                transportFetchOperation,
+                threadPool.executor(ThreadPool.Names.SUGGEST),
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java]: You could use `Executors.new...` here instead of creating a ThreadPool (ThreadPool creates a dozen of threads which is a waste for the test)

File: sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    public void after() throws Exception {
+        threadPool.shutdown();
+        threadPool.awaitTermination(1, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFetchSize() throws Throwable {
+        CollectingRowReceiver rowReceiver = new CollectingRowReceiver();
+
+        TransportFetchOperation transportFetchOperation = mock(TransportFetchOperation.class);
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java]: Would it be possible to implement/mock the interface (FetchOperation) instead of the concrete implementation?

File: sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            indexToTable
+        );
+    }
+
+    private List<Symbol> buildOutputSymbols() {
+        List<Symbol> outputSymbols = new ArrayList<>(2);
+
+        InputColumn inputColumn = new InputColumn(0);
+        ReferenceIdent referenceIdent = new ReferenceIdent(USER_TABLE_IDENT, "id");
+        ReferenceInfo referenceInfo = new ReferenceInfo(referenceIdent,
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java]: Just FYI: To create Reference's there would also be `SqlExpressions.asSymbol(...)` - but I'm not sure if it makes sense to use it here

File: sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java (hunk 1 above)
Comment (matriv) [inline sql/src/main/java/io/crate/executor/transport/TransportFetchNodeAction.java]: Normally we don't send requests where there is nothing to fetch and isCloseContext is false:
Check code if FetchProjector line 182:

`
if (toFetch.isEmpty() && !isLast) {
     remainingRequests.decrementAndGet();
 } else {
     final String nodeId = entry.getKey();
     ListenableFuture<IntObjectMap<? extends Bucket>> future = fetchOperation.fetch(nodeId, toFetch, isLast);
`

File: sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        int fetchSize = 3;
+        FetchProjector fetchProjector = prepareFetchProjector(fetchSize, rowReceiver, transportFetchOperation, upstream);
+
+        IntObjectMap<Bucket> intObjectMap = new IntObjectHashMap<>(2);
+        Collection<Object[]> rows = new ArrayList<>(10);
+
+        int i;
+        for (i = 1; i <= 10; i++) {
+            Row row = new Row1((long)i);
+            rows.add(new Object[]{row});
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/operation/projectors/FetchProjectorTest.java]: This seems a bit odd.  Either rows should contain just Row instances or Object arrays that contain the values, but not other rows.

You could use:
```
Row row = new Row1(i);
rows.add(row.materialize());
```

**Comment by mfussenegger** [comment]
Would be great to have separate unit tests for the RowMergers and PageDownstream fixes

**Comment by mfussenegger** [comment]
lgtm
